### ISSUE TITLE
[BUG]: Rust: claims that `main` function is unused even when running the binary

### ISSUE THREAD

**Issue by RalfJung** [compiler-explorer/compiler-explorer#5350]
### Describe the bug

Even when building a binary, godbolt claims that `main` is unused.

### Steps to reproduce

1. Visit rust.godbolt.org
2. enter
```rust
fn main() {
    println!("Hello world!");
}
```
3. select "Execute the code"

This leads to a warning being shown in the output window:
```
warning: function `main` is never used
 --> <source>:1:4
  |
1 | fn main() {
  |    ^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Expected behavior

The code should run and there should be no warnings.

### Reproduction link

https://rust.godbolt.org/z/1aajhn6cc

### Screenshots

Not applicable

**Comment by partouf** [comment]
The main fn needs to be pub

**Comment by RalfJung** [comment]
No, it doesn't. There is [no warning here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ee20c78137b6276200e460c68560e92a).

**Comment by RalfJung** [comment]
My hypothesis would be that godbolt builds the code once with `--crate-type lib`, and that's when the warning is emitted, and then builds the code again with `--crate-type bin`, which is used to produce the binary that runs. Or something like that.

**Comment by partouf** [comment]
> My hypothesis would be that godbolt builds the code once with `--crate-type lib`, and that's when the warning is emitted, and then builds the code again with `--crate-type bin`, which is used to produce the binary that runs. Or something like that.

Correct

**Comment by RalfJung** [comment]
It should then ideally avoid building binary crates with `--crate-type lib`, that's now how they are meant to be built.

**Comment by partouf** [comment]
> It should then ideally avoid building binary crates with `--crate-type lib`, that's now how they are meant to be built.

changing the type is being discussed here https://github.com/compiler-explorer/compiler-explorer/issues/4910
